File: src/main/com/dune/lang/parser/DuneParserDefinition.java
package com.dune.lang.parser;

import com.dune.DuneLanguage;
import com.dune.ide.files.DuneFile;
import com.dune.lang.core.psi.impl.DuneTypes;
import com.dune.lang.lexer.DuneLexer;
import com.intellij.lang.ASTNode;
import com.intellij.lang.ParserDefinition;
import com.intellij.lang.PsiParser;
import com.intellij.lexer.FlexAdapter;
import com.intellij.lexer.Lexer;
import com.intellij.openapi.project.Project;
import com.intellij.psi.FileViewProvider;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.TokenType;
import com.intellij.psi.tree.IFileElementType;
import com.intellij.psi.tree.TokenSet;
import org.jetbrains.annotations.NotNull;

public class DuneParserDefinition implements ParserDefinition {
    private static final TokenSet WHITE_SPACES = TokenSet.create(TokenType.WHITE_SPACE);
    private static final TokenSet COMMENTS = TokenSet.create(DuneTypes.COMMENT);
    private static final TokenSet STRINGS = TokenSet.create(DuneTypes.STRING);

    private static final IFileElementType FILE = new IFileElementType(DuneLanguage.INSTANCE);

    @NotNull
    @Override
    public Lexer createLexer(Project project) {
        return new FlexAdapter(new DuneLexer(null));
    }

    @NotNull
    public TokenSet getWhitespaceTokens() {
        return WHITE_SPACES;
    }

    @NotNull
    public TokenSet getCommentTokens() {
        return COMMENTS;
    }

    @NotNull
    public TokenSet getStringLiteralElements() {
        return STRINGS;
    }

    @NotNull
    public PsiParser createParser(final Project project) {
        return new DuneParser();
    }

    @NotNull
    @Override
    public IFileElementType getFileNodeType() {
        return FILE;
    }

    @NotNull
    public PsiFile createFile(@NotNull FileViewProvider viewProvider) {
        return new DuneFile(viewProvider);
    }

    @NotNull
    public SpaceRequirements spaceExistenceTypeBetweenTokens(ASTNode left, ASTNode right) {
        return SpaceRequirements.MAY;
    }

    @NotNull
    public PsiElement createElement(@NotNull ASTNode node) {
        return DunePsiElementFactory.createElement(node);
    }
}

File: src/main/kotlin/com/dune/language/parser/DuneParserDefinition.kt
package com.dune.language.parser

import com.dune.language.DuneLanguage
import com.dune.language.lexer.DuneLexerAdapter
import com.dune.language.psi.DuneTypes
import com.dune.language.psi.files.DuneFile
import com.intellij.lang.ASTNode
import com.intellij.lang.ParserDefinition
import com.intellij.lang.PsiParser
import com.intellij.lexer.Lexer
import com.intellij.openapi.project.Project
import com.intellij.psi.FileViewProvider
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.tree.IFileElementType
import com.intellij.psi.tree.TokenSet

class DuneParserDefinition : ParserDefinition {
    override fun createLexer(project: Project?): Lexer = DuneLexerAdapter()

    override fun createParser(project: Project?): PsiParser = DuneParser()

    override fun getFileNodeType(): IFileElementType = FILE

    override fun getCommentTokens(): TokenSet = COMMENT_TOKENS

    override fun getStringLiteralElements(): TokenSet = STRING_TOKENS

    override fun createElement(node: ASTNode): PsiElement = DuneTypes.Factory.createElement(node.elementType)

    override fun createFile(viewProvider: FileViewProvider): PsiFile = DuneFile(viewProvider)

    companion object Utils {
        val FILE = IFileElementType(DuneLanguage)
        val COMMENT_TOKENS = TokenSet.create(DuneTypes.COMMENT)
        val STRING_TOKENS = TokenSet.create(DuneTypes.STRING_VALUE)
    }
}